### DISCUSSION TITLE
read H264 Byte stream

### DISCUSSION THREAD

**Discussion by ashif-ismail** [kokorin/Jaffree#344] [Q&A]
Hello, some one can help regarding my issue below,

i receive H264 bytestream from a websocket in my spring boot application, i need this h264 to be sent to angular frontend. i tried using video tag which does not work.

any suggestions pls.

**Comment by kokorin** [comment]
> i tried using video tag which does not work

Could you clarify on that? I'm not sure what tag you meant.

First of all you need to verify that bytes received through WebSocket can be read by ffmpeg. Check [PipeInput](https://github.com/kokorin/Jaffree#supplying-and-consuming-data-with-inputstream-and-outputstream) example for that.

The most optimal way IMHO to resend [video stream to a browser is HLS](https://github.com/kokorin/Jaffree#live-stream-re-streaming-hls).
HLS index (m3u8) and chunks (ts) can be served by Spring as files from local disk.

**Reply by ashif-ismail** [reply]
This is what i have 

1)Byte stream of h264 from a socket (in spring boot backend)
2)a frontend angular

i need a way to transmit the h264 bytes to frontend, may be ismv

i tried HLS approch which works fine, but it has a delay about 30s which is not acceptable

**Reply by ashif-ismail** [reply]
i have used the following code to write HLS chunks to disk, which is taking 30s to start the stream 

`FFmpeg.atPath().addArguments("-rtsp_transport", "tcp")
                    .addArguments("-i", streamingLink).addArguments("-f", "hls")
                    .addArguments("-vsync", "0").addArgument("-copyts")
                    .addArguments("-vcodec", "copy")
                    .addArguments("-b:a", "96k")
                    .addArguments("-acodec", "copy")
                    .addArguments("-movflags", "frag_keyframe+empty_moov")
                    .addArguments("-hls_flags", "delete_segments+append_list")
                    .addArguments("-hls_time", "5").addArguments("-hls_list_size", "5")
                    .addArgument(requestingCameraPath.getPath() + fileSeparator + "index.m3u8").executeAsync();`

any chance for improving the performance here

**Reply by kokorin** [reply]
> i tried HLS approch which works fine, but it has a delay about 30s which is not acceptable

Time lag depends on many factors: 
1. do you re-encode or keep original stream encoding?
2. what is configured segment length?
3. what is segment length that you see in M3U8 index?

If you need to have minimal delay and don't need to support to much viewers, probably, you can just implement cyclic buffer (filled from source WebSocket) and resend buffered video through WebSockets to clients.
The last option doesn't require Jaffree or ffmpeg.

**Reply by ashif-ismail** [reply]
the application that i'm building is just for an internal usage, very minimally used by countable users (less than 5 at a time)
please guide through your points as im new to media encoding and similar stuff

thanks

**Reply by ashif-ismail** [reply]
how do i know this  ?

do you re-encode or keep original stream encoding?

**Reply by ashif-ismail** [reply]
reagarding your last point, could you please explain more... i did a small poc using SSEEmmitter in spring boot that resends bytestream to client through API. but the problem is .. how do i play H264 bytream in HTML frontend ?

**Reply by kokorin** [reply]
> please guide through your points as im new to media encoding and similar stuff

I'm trying, but you don't answer my questions

> how do i know this ?

`addArguments("-vcodec", "copy")` this means you copy codec and hence keep original keyframes.  So it keyframe is inserted every 10 seconds - the average delay will be not less then 10 seconds and should be not higher then 20 seconds.
So if you want to go with HLS you either need to lower keyrame interval to 1-2 seconds. or to re-encode video with explicitly set keyframe interval.

**Reply by ashif-ismail** [reply]
ok ,

1. How do i lower the keyframe interval (could you please suggest from my above command)
2. could you please breif on sending the bytestream directly to client as per your third point in previous answer

thanks

**Reply by kokorin** [reply]
> How do i lower the keyframe interval (could you please suggest from my above command)

Please, try to find ffmpeg documentation. 

> could you please breif on sending the bytestream directly to client as per your third point in previous answer

Please, try to find Spring documentation on that.

**Reply by ashif-ismail** [reply]
a link for the second point (spring doc) would be appreciated, i have managed to reduce the keyframe interval for ffmpeg.
